Add probability-based stepping to a node in network

wenzeslaus opened the pull request:
This adds step (i.e., jump) from one node to another based on probability assigned to an edge.

The code it uses is completely different, so a new function _step_ is now available alongside of the existing _travel_ function.

Node to go to is selected based on edge probability in the new _step_ function. Ignoring of nodes, implemented for the _travel_ function, is not implemented for the _step_ function because that would mean recomputing the probabilities.

petrasovaa commented on include/pops/network.hpp:
```diff
@@ -731,15 +813,26 @@ class Network
             }
-            if (label == "cost") {
+            if (label == "probability") {
+                if (has_cost) {
+                    std::runtime_error(
+                        "The cost column must be after the probability column");
+                }
```
I wonder if this is not redundant when you test for column number below anyway.

wenzeslaus replied:
Yes, captured by the test below, too, but provides more a relevant message to the user. I added comment exampling that. Also added tests and discovered a missing throw there (we don't have automatic checks implemented to catch that). Done in c8c9949.

petrasovaa reviewed (changes requested):
It looks like recomputing the probabilities would need to be added, it looks like with highly probable edges, you would be going there and back a lot, no?

wenzeslaus commented on include/pops/network.hpp:
```diff
@@ -487,6 +500,36 @@ class Network
+    std::tuple<int, int> step(
+        RasterIndex row, RasterIndex col, Generator& generator, int num_steps = 1) const
+    {
+        auto node_id = get_random_node_at(row, col, generator);
```
Right. This was documented in travel function code. Now documented in doc for both functions. 0f45f0c

wenzeslaus commented on include/pops/network.hpp:
```diff
@@ -487,6 +500,36 @@ class Network
+     *
+     * If *num_steps* is greater than 1, multiple steps are perfomed and the last node
+     * is returned. In each node, the probability to picking a specific connection are
+     * either determined by the provided edge probabilities or are equal. Consequently,
```
Thanks. Done in 151f4b3.

wenzeslaus commented on include/pops/network.hpp:
```diff
@@ -680,6 +724,42 @@ class Network
+    static double probability_from_text(const std::string& text)
+    {
+        try {
+            return std::stod(text);
```
Good point. 04744f9 adds the check and tests for the good and bad cases.

wenzeslaus replied:
Since f2d11d54a86a011c6b51f390e97c894c0230aa27 only `>1` is enforced. Any other range is allowed and meaning is up to the user thanks to the underlying std::discrete_distribution function.

ChrisJones687 reviewed (changes requested):
Overall this looks great just my one comment on whether the variable is bool or double.

wenzeslaus commented:
> it looks like with highly probable edges, you would be going there and back a lot, no?

That's correct, but the kernel actually always does only one step because that was the original goal. I ended up adding the multiple steps because it was simple addition and useful for testing. Multiple steps would still work well for network without edge probabilities.

wenzeslaus commented:
I checked again the probabilities for edges and the reason why I didn't originally address the ranges is that the C++ function (std::discrete_distribution: [cppreference.com](https://en.cppreference.com/w/cpp/numeric/random/discrete_distribution), [cplusplus.com](https://www.cplusplus.com/reference/random/discrete_distribution/)) I use does not care about the actual numbers. It uses them as weights, i.e., they can be in any range and their sum does not need to be 1 or 100.

Now, we can check the probabilities or not check them at all.
- If we don't check them, they are just relative weights. You can have 1, 2, 3 (sum 6) for three edges from one node, or 10, 70, 20 (sum 100), or 0.1, 0.7, 0.2 (sum 1) or 0.1, 0.2, 0.3 (sum 0.6). It is up to the user to ensure that it makes sense, but it might be convenient since likely the edge probability is based on some weights anyway. (We would still check for negative values which is one of the tests I added based on Anna's comment.)
- We can check the value range. This is what I added based on Anna's comment, so 0.1, 0.7, 0.2 (sum 1) and 0.1, 0.2, 0.3 (sum 0.6) are allowed, but not the others.
- Additionally, we can check the sum for one node. I need to have the values in one list anyway while loading the network. However, checking that the sum is exactly 1 (or 100) is tricky in floating point numbers, so I would rather avoid that if the benefit is unclear. When you generate the probabilities for the edges, that's the best place to ensure that they are equal to 1 (or 100). The network does not care given the underlying C++ function, so the check would be unrelated to the network itself.

wenzeslaus commented:
Config for the network is now `std::string network_type` with values `"travel"`, `"snap"`, and `"step"`.

The three options are:
1. Go through the network pixel by pixel using the cost (distance), stop at any pixel.
2. Go through the network pixel by pixel using the cost (distance), stop at nodes.
3. Go through the network from node to node. 

There is actually not much difference between 2 and 3 spatially or in the implementation, but the big difference is that edge cost is considered in 2 and not in 3.

The option 3 is different from 1 and 2 in use of probability for edges (picking which edge to go from a node). This is now completely missing in 1 and 2, but it can be added. At the same time, unique edge probabilities are optional for 3 and the use of the unique edge probabilities (as opposed to equal probabilities) is determined by the probability column being present in the network file rather than by the configuration, so 3 is not defined by use of edge probabilities.

These are options I came up with:
* travel, snap, step (current, step and snap are too similar)
* travel, snap, jump (jump is the only other thing I came up with)
* walk, snap, jump (walk seems to pair well with jump)
* cost, cost-snap, jump (jump with cost would be like cost with snapping ignoring the edge probabilities difference)
* pixel, node, jump (getting to the interaction with the physical world, but node and jump actually interact in the same way)
* cell, node, jump (PoPS Core mostly uses cell, not pixel)
* spatial, nodes, jump (the spatial part is what makes it different from classic networks in network modeling or GIS)
* open carriage, closed carriage, teleport (based on the mode of transportation, but trying to be abstract)
* open carriage, closed carriage, transporter (Star Trek)
* open carriage, closed carriage, wormhole (Stargate)
* open carriage, closed carriage, plane (less science fiction, but the two words per option are sub-optimal)
* chariot, wagon, transporter (too far?)

For this PR, I would like to leave it as it is in case it includes changing travel and snap because that's really outside of this PR.

wenzeslaus commented:
@ChrisJones687 @petrasovaa just approve this if you agree with latest changes. No need to decide the naming now.

petrasovaa commented on tests/test_network.cpp:
```diff
@@ -282,6 +282,187 @@ int test_cost_network()
+        return 1;
+    }
+    catch (const std::runtime_error& error) {
+        // All is good. We expect and exception.
```
```suggestion
        // All is good. We expect an exception.
```

wenzeslaus replied:
Thanks. Done in 456b10f.
